hledger-lib/Hledger/Data/Transaction.hs

```diff
@@ -551,38 +551,37 @@ inferBalancingPrices t@Transaction{tpostings=ps} = t{tpostings=ps'}
 -- price to the posting, if and as appropriate for the given transaction and
 -- posting type (real or balanced virtual).
 priceInferrerFor :: Transaction -> PostingType -> (Posting -> Posting)
-priceInferrerFor t pt = inferprice
+priceInferrerFor t pt = maybe id inferprice $ inferFromAndTo sumamounts
   where
-    postings       = filter ((==pt).ptype) $ tpostings t
-    pamounts       = concatMap (amounts . pamount) postings
-    pcommodities   = map acommodity pamounts
-    sumamounts     = amounts $ sumPostings postings  -- sum normalises to one amount per commodity & price
-    sumcommodities = map acommodity sumamounts
-    sumprices      = filter (/=Nothing) $ map aprice sumamounts
-    caninferprices = length sumcommodities == 2 && null sumprices
-
-    inferprice p@Posting{pamount=Mixed [a]}
-      | caninferprices && ptype p == pt && acommodity a == fromcommodity
-        = p{pamount=mixedAmount $ a{aprice=Just conversionprice}, poriginal=Just $ originalPosting p}
+    postings     = filter ((==pt).ptype) $ tpostings t
+    pcommodities = map acommodity $ concatMap (amounts . pamount) postings
+    sumamounts   = amounts $ sumPostings postings  -- sum normalises to one amount per commodity & price
+    noprices     = all (isNothing . aprice) sumamounts
+
+    -- Determine the amounts we are inferring from and to based on the order of commodities in pcommodities
```

Comment:
Does this change any user-visible behaviour ? 

More generally, priceInferrerFor looks a bit hard to understand - my fault, but since it's fresh in your mind if you see any additional comments that would help please feel free to add.

Reply:
It should not change any user-visible behaviour. I made the changes one replacement at a time, checking that all tests passed at each stage. I agree that the function is a bit hard to understand, and I'm happy to add more comments.

Reply:
I've added some more comments. Let me know if there's anything that you think could be clarified further.